File: 1516-the-k-th-lexicographical-string-of-all-happy-strings-of-length-n/solution.cpp
class Solution {
public:
    int count = 0;
    string result = "";
    
    void generate(int n, string current, char lastChar, int &k) {
        if (current.size() == n) {
            count++;
            if (count == k) result = current;
            return;
        }
        
        for (char c : {'a', 'b', 'c'}) {
            if (c != lastChar) {
                generate(n, current + c, c, k);
                if (!result.empty()) return;
            }
        }
    }
    
    string getHappyString(int n, int k) {
        generate(n, "", '#', k);
        return result;
    }
};

File: 1516-the-k-th-lexicographical-string-of-all-happy-strings-of-length-n/solution.java
class Solution {
    private int cnt = 0;
    private String result = "";

    public String getHappyString(int n, int k) {
        if (k > 3 * (1 << (n - 1))) {
            return "";
        }
        helper(n, new StringBuilder(), '#', k);
        return result;
    }

    private void helper(int n, StringBuilder sb, char lastChar, int k) {
        if (result.length() > 0) {
            return;
        }

        if (sb.length() == n) {
            cnt++;
            if (cnt == k) {
                result = sb.toString();
            }
            return;
        }

        for (char c : new char[]{'a', 'b', 'c'}) {
            if (c != lastChar) {
                sb.append(c);
                helper(n, sb, c, k);
                if (result.length() > 0) {
                    return;
                }
                sb.deleteCharAt(sb.length() - 1); 
            }
        }
    }
}

